### ISSUE TITLE
Question on incremental training and a suggestion

### ISSUE THREAD

**Issue by mohammadm1985** [MaartenGr/BERTopic#1429]
Hi, Thank you so much for creating such a cool and interactive and highly modular technique in Topic Modeling.

I have a question and a recommendation.

Regarding the question, I'm curious if there are any options for online training while still using HDBSCAN and UMAP. These methods offer powerful concepts for dimensionality reduction and hierarchical density-based clustering, and I would like to retain their benefits. Other methods tend to lack support for noise detection since they don't include a density component. I'm interested in your suggestions for online training that can preserve these outstanding components.

Additionally, I'd like to recommend using the following function to wrap the document text for hovertext:

python
Copy code
def insert_newlines(text, every=20):
    words = text.split()
    lines = [' '.join(words[i:i+every]) for i in range(0, len(words), every)]
    return '<br>'.join(lines)

By applying this function to selection["doc"], the document text can be made more readable with hovertext in document plots.
I've tested the function with every = 20, and it works effectively with documents of any length.

Furthermore, I have developed several evaluation metrics for Bertopics. The source code was originally from OCTIS, but I have completely adapted it to work with the vocabularies used in the Bertopic vectorizer model's vocabulary. During the adaptation, I discovered and fixed a few logical bugs in the OCTIC main functions, including normalization requirements and other logical issues, making them optimized for Bertopic. I've also created some distance metrics to evaluate the model by shuffling the topic words and bootstrapping from the distribution of the words. The results have been very promising. If you're interested, I'd be delighted to collaborate with you on this project and incorporate these enhancements into the main function. Please let me know your preferred method of communication, and I will reach out to you accordingly. You can contact me via email at mmnejad@terpmail.umd.edu.

**Comment by mohammadm1985** [comment]
I would like to add that I am working on clustering customer complaints and as time passes new complaint trends might appear and I might need to consider new clusters. If I do the training again the clustering will change slightly due to randomized processes in UMAP reduction steps and the distribution of topic documents may change as well. The number of clusters can change slightly as well. All this leads to instability. Therefore, I would like to be able to maintain the original structure of clusters with partial training. Methods other than HDBSCAN that support partial training, do not support noise detection. I would like to know if there is a reliable way to add new clusters without disturbing the distribution of documents in previous clusters while using the HDBSCAN.

**Comment by MaartenGr** [comment]
> Regarding the question, I'm curious if there are any options for online training while still using HDBSCAN and UMAP. These methods offer powerful concepts for dimensionality reduction and hierarchical density-based clustering, and I would like to retain their benefits. Other methods tend to lack support for noise detection since they don't include a density component. I'm interested in your suggestions for online training that can preserve these outstanding components.

There is no support for partial training with HDBSCAN clustering at the moment, you would have to add that yourself. With respect to UMAP, there is a method of doing so albeit it can be a bit tricky to implement. I believe there are density-based clustering algorithms in River but I am not entirely sure.

> I would like to add that I am working on clustering customer complaints and as time passes new complaint trends might appear and I might need to consider new clusters. If I do the training again the clustering will change slightly due to randomized processes in UMAP reduction steps and the distribution of topic documents may change as well. The number of clusters can change slightly as well. All this leads to instability. Therefore, I would like to be able to maintain the original structure of clusters with partial training. Methods other than HDBSCAN that support partial training, do not support noise detection. I would like to know if there is a reliable way to add new clusters without disturbing the distribution of documents in previous clusters while using the HDBSCAN.

This is currently not possible since HDBSCAN does not support partial training. However, it might be possible to cluster the newly found outlier topics and add those to the training data. As a result, you will have a list of existing topics with a bunch of new topics. You can then pass those on to [manual BERTopic](https://maartengr.github.io/BERTopic/getting_started/manual/manual.html) to create a topic model that has your existing topics together with a bunch of new ones. 

> Additionally, I'd like to recommend using the following function to wrap the document text for hovertext:

I have to look into this a bit more. It was suggested before to split the text up but it resulted in very long messages instead of very wide messages. Also, splitting by spaces does not work for many non-western languages and it would have to be a language-independent process.

**Comment by mohammadm1985** [comment]
Thank you so much for the guidance. Would you please keep the tissue open for a week? I might have some questions regarding this after having more experience.

I think you can define an attribute that lets the user choose the type of language and break the text accordingly. Right now, the text is shown in one line, and for long texts you won't even be able to see the text and the main part of the semantic might be hidden somewhere outside the plot borders.

**Comment by mohammadm1985** [comment]
I am also thinking, maybe I can fix the random seed and run the model for 1 year and then predict for a few months ahead. and then use the same seed to train the model again after a few months. It should pick the previous topics and potentially new topics that emerge as well.

**Comment by mohammadm1985** [comment]
This one should work with every language:

def insert_newlines(text, every=20, lan_type = "sep"): # lan_type can be symb for languages like Chinese and Jap
    words = text.split()
    lines = [' '.join(words[i:i+every]) for i in range(0, len(words), every)]
    if lan_type == 'symb':
        _unicode_chr_splitter = re.compile( '(?s)((?:[\ud800-\udbff][\udc00-\udfff])|.)' ).split
        words = [ chr for chr in _unicode_chr_splitter( text ) if chr ]
        lines = [''.join(words[i:i+every]) for i in range(0, len(words), every)]
    return ' \n '.join(lines)

The code works for other languages that are right to left like Arabic.

**Comment by MaartenGr** [comment]
> This one should work with every language:

I believe there is another issue where there is a bit more info about splitting text for visualization purposes but generally, I would prefer using the tokenizer that is being used to tokenized the documents. That way, users do not have to additionally specify languages as they already did that before when training the model. 

> Thank you so much for the guidance. Would you please keep the tissue open for a week?

Sure, I'll keep it open.

**Comment by mohammadm1985** [comment]
> I believe there is another issue where there is a bit more info about splitting text for visualization purposes but generally, I would prefer using the tokenizer that is being used to tokenized the documents. That way, users do not have to additionally specify languages as they already did that before when training the model.

Yeah... It is smart

**Comment by mohammadm1985** [comment]
@MaartenGr 
I decided to train the model for a year and predict for a few months ahead and then update the model and keep doing this interval. However, When I am trying to predict the new data, for most of the cases they fall into -1 category. I am pretty sure that the text context has not changed a lot. Originally 30 % of the data was captured as outliers but for prediction data 90% are outliers. I took the max from probability calculated for new docs, but I am still getting inconsistent results. Comparing the context with the assigned documents it is not relevant. The training set is well-formed and pretty consistent though.

Also, I was wondering if there is a visualization tool for new documents (prediction set). I tried using the visualize_document() func, but the internal function reads the documents from model_topic.topics_ which returns a dimension error when using new data as docs.

**Comment by MaartenGr** [comment]
It depends on the underlying cluster model that you used but I assume you are using HDBSCAN. That model tends to assign documents to -1 using its prediction task. What you can also do is use the methods from `.reduce_outliers` to assign outliers to topics. I would advise using the embeddings strategy. 

> Also, I was wondering if there is a visualization tool for new documents (prediction set). I tried using the visualize_document() func, but the internal function reads the documents from model_topic.topics_ which returns a dimension error when using new data as docs.

It is currently only implemented for documents that the model was trained on since those represent the model.

**Comment by mohammadm1985** [comment]
Thank you for the clarification and all your assistance. Based on your inputs, it seems that HDBSCAN may not be the most suitable algorithm for prediction purposes. In light of this, I would appreciate your recommendation for an alternative clustering algorithm that would better suit the prediction needs.

Additionally, I'm curious if it is possible to apply the `.reduce_outliers` function to the new prediction data.

**Comment by MaartenGr** [comment]
> Based on your inputs, it seems that HDBSCAN may not be the most suitable algorithm for prediction purposes. In light of this, I would appreciate your recommendation for an alternative clustering algorithm that would better suit the prediction needs.

I would advise still using HDBSCAN for the initial clustering and then do the predictions by comparing the embeddings from the input documents with the topic embeddings through cosine similarity. It is similar to `.reduce_outliers` which you can also use for new prediction data. 

> I noticed that I can generate topic distribution for any new document without using the transform function. Can I use that to assign the new documents to a topic with the highest probability?

Yes, that is also a possibility. Do note though that it is a different procedure from the cosine similarity method I explained above. The former is based on classifying entire documents whereas the latter is based on classifying token sets and aggregating those. Here, I would also advise using embeddings instead of the c-TF-IDF strategy.

**Comment by mohammadm1985** [comment]
Thank you!

Would you also suggest using parametric UMAP which is deterministic in terms of predictions and is not batch dependent? The algorithm uses the same component to construct the graph out of the data but uses a neural network encoder to reduce the dimensionality while preserving the structure of the graph. This will allow bypassing the limitations of UMAP prediction. (UMAP generates different results for prediction due to its stochastic nature when batch size is different from training batch size, which is always the case - for example, I have 4K training data but 500 prediction data).

**Comment by mohammadm1985** [comment]
Thank you for your time!

> ... do the predictions by comparing the embeddings from the input documents with the topic embeddings through cosine similarity.

This is quite confusing to me. I don't know whether to use `.reduce_outliers` on prediction data or write my own code to compare `1-cosine` and set a threshold to pick a topic for a document. 
> It is because UMAP stochastically changes the results when using transform with new data...

Would you also suggest using parametric UMAP which is deterministic in terms of predictions and is not batch dependent? The algorithm uses the same component to construct the graph out of the data but uses a neural network encoder to reduce the dimensionality while preserving the structure of the graph. This will allow bypassing the limitations of UMAP prediction. (UMAP generates different results for prediction due to its stochastic nature when batch size is different from training batch size, which is always the case - for example, I have 4K training data but 500 prediction data).

**Comment by MaartenGr** [comment]
> This is quite confusing to me. I don't know whether to use .reduce_outliers on prediction data or write my own code to compare 1-cosine and set a threshold to pick a topic for a document.

You should be able to use `.reduce_outliers` for this purpose. 1-cosine is the same method. 

> Would you also suggest using parametric UMAP which is deterministic in terms of predictions and is not batch dependent? The algorithm uses the same component to construct the graph out of the data but uses a neural network encoder to reduce the dimensionality while preserving the structure of the graph. This will allow bypassing the limitations of UMAP prediction. (UMAP generates different results for prediction due to its stochastic nature when batch size is different from training batch size, which is always the case - for example, I have 4K training data but 500 prediction data).

I have not used parametric UMAP myself before but it might be worth trying out. 

> One more question: is there a way that I can change clusterer without retraining the bertopic? Let's say I have used a dimension reduction technique and it took me some time to train it and I wan to change the clustering alg parameters, but not retrain the whole model, only see the results with the new clusterer and maybe run representation model for it. Is this possible? Is there a specific function to call?

There are multiple ways to approach this. You can skip over dimensionality reduction through [this guide](https://maartengr.github.io/BERTopic/getting_started/dim_reduction/dim_reduction.html#skip-dimensionality-reduction). You could also use `.update_topics` to only update the topic representations and skip over re-fitting entirely. Finally, you can use [manual topic modeling](https://maartengr.github.io/BERTopic/getting_started/manual/manual.html) if you already have clusters.
